src/dynamodb-mcp-server/awslabs/dynamodb_mcp_server/model_validation_utils.py

```diff
@@ -0,0 +1,818 @@
+        'ps',
+        '--format',
+        '{{.Ports}}',
+        '-f',
+        f'name={DynamoDBLocalConfig.CONTAINER_NAME}',
+    ]
+    result = _run_subprocess_safely(ports_cmd)
+
+    if result and result.stdout.strip():
+        return _parse_container_port(result.stdout.strip())
+    return None
+
+
+def _extract_port_from_cmdline(cmdline: list) -> Optional[int]:
+    """Extract port number from Java command line arguments.
+
+    Args:
+        cmdline: List of command line arguments from process
+
+    Returns:
+        Optional[int]: Port number if found and valid, None otherwise
+    """
+    for i, arg in enumerate(cmdline):
+        if arg == '-port' and i + 1 < len(cmdline):
+            try:
+                return int(cmdline[i + 1])
+            except ValueError:
+                return None
+    return None
+
+
+def _safe_extract_members(members):
+    """Filter tar members to prevent path traversal attacks.
+
+    Args:
+        members: Iterable of tar members
+
+    Yields:
+        Safe tar members that don't contain path traversal sequences
+    """
+    for member in members:
+        if os.path.isabs(member.name) or '..' in member.name:
+            continue
+        yield member
+
+
+def _get_dynamodb_local_paths() -> tuple[str, str, str]:
+    """Get paths for DynamoDB Local artifacts."""
+    dynamodb_dir = os.path.join(tempfile.gettempdir(), DynamoDBLocalConfig.TEMP_DIR_NAME)
+    jar_path = os.path.join(dynamodb_dir, 'DynamoDBLocal.jar')
+    lib_path = os.path.join(dynamodb_dir, 'DynamoDBLocal_lib')
+    return dynamodb_dir, jar_path, lib_path
+
+
+def _validate_download_url(url: str) -> None:
+    """Validate download URL to prevent security issues.
+
+    Args:
+        url: URL to validate
+
+    Raises:
+        ValueError: If URL is not safe to use
+    """
+    if not url.startswith('https://'):
+        raise ValueError(f'Only HTTPS URLs are allowed, got: {url}')
+
+    # Only allow AWS CloudFront domains for DynamoDB Local downloads
+    allowed_domains = ['d1ni2b6xgvw0s0.cloudfront.net']
+    from urllib.parse import urlparse
+
+    parsed = urlparse(url)
+    if parsed.hostname not in allowed_domains:
+        raise ValueError(f'URL domain not allowed: {parsed.hostname}')
+
+
+def _download_and_extract_jar(dynamodb_dir: str, jar_path: str, lib_path: str) -> None:
+    """Download and extract DynamoDB Local JAR."""
+    tar_path = os.path.join(dynamodb_dir, DynamoDBLocalConfig.TAR_FILENAME)
+
+    logger.info('Downloading DynamoDB Local...')
+
+    try:
+        # Validate URL before download
+        _validate_download_url(DynamoDBLocalConfig.DOWNLOAD_URL)
+
+        # Download with timeout
+        with urllib.request.urlopen(  # nosec B310
+            DynamoDBLocalConfig.DOWNLOAD_URL, timeout=DynamoDBLocalConfig.DOWNLOAD_TIMEOUT
+        ) as response:
+            with open(tar_path, 'wb') as f:
+                f.write(response.read())
+
+        # Extract the tar.gz file safely
```

Comment:
If this will fail without `DynamoDBLocal.jar`, would testing it before extracting help?